File: patches/fix_buddy_ble_linux.py
#!/usr/bin/env python3
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: python
"""
Enable Hardware Buddy (Nibblet BLE device) on Linux.

The Buddy feature communicates with a Nibblet (M5StickC Plus) over BLE using
the Nordic UART Service (NUS). The BLE transport runs entirely through Web
Bluetooth in Electron's renderer — no native code involved. On macOS it works
natively; on Linux Electron uses BlueZ/D-Bus (requires bluez package).

Two issues on Linux:
  A) GrowthBook flag 2358734848 gates the feature — force it on.
  B) Race condition: the buddy_window renderer calls reportState("ready")
     during preload, before the async BLE bridge init completes and
     setImplementation() registers the real handlers. Register early stubs
     at app.on("ready") so the renderer doesn't hit missing handlers.
     The real setImplementation() calls removeHandler() before handle(),
     so the stubs are cleanly replaced once the bridge initializes.

Prerequisites (Linux): bluez package for BLE support in Chromium/Electron.

Usage: python3 fix_buddy_ble_linux.py <path_to_index.js>
"""

import sys
import os
import re


EXPECTED_PATCHES = 2


def extract_eipc_uuid(content):
    """Extract the eipc UUID from the file content."""
    m = re.search(
        rb"\$eipc_message\$_([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})",
        content,
    )
    if m:
        return m.group(1).decode("utf-8")
    return None


def build_early_stubs_js(eipc_prefix):
    """Early no-op IPC handlers replaced when setImplementation() runs."""
    buddy = eipc_prefix + "claude.buddy_$_Buddy_$_"
    ble = eipc_prefix + "claude.buddy_$_BuddyBleTransport_$_"

    return (
        'if(process.platform==="linux"){'
        'const _bi=require("electron").ipcMain;'
        # BuddyBleTransport — called by renderer preload before bridge init
        f'_bi.handle("{ble}reportState",()=>{{}});'
        f'_bi.handle("{ble}rx",()=>{{}});'
        f'_bi.handle("{ble}log",()=>{{}});'
        # Buddy — called by buddy_window renderer
        f'_bi.handle("{buddy}status",()=>({{connected:false,error:null,paired:false}}));'
        f'_bi.handle("{buddy}deviceStatus",()=>null);'
        f'_bi.handle("{buddy}setName",()=>null);'
        f'_bi.handle("{buddy}pairDevice",()=>"");'
        f'_bi.handle("{buddy}scanDevices",()=>[]);'
        f'_bi.handle("{buddy}pickDevice",()=>false);'
        f'_bi.handle("{buddy}cancelScan",()=>{{}});'
        f'_bi.handle("{buddy}forgetDevice",()=>{{}});'
        f'_bi.handle("{buddy}pickFolder",()=>null);'
        f'_bi.handle("{buddy}preview",()=>null);'
        f'_bi.handle("{buddy}install",()=>{{}});'
        "}"
    )


def patch_buddy_ble(filepath):
    """Enable Buddy BLE on Linux: force flag + early stubs."""

    print("=== Patch: fix_buddy_ble_linux ===")
    print(f"  Target: {filepath}")

    if not os.path.exists(filepath):
        print(f"  [FAIL] File not found: {filepath}")
        return False

    with open(filepath, "rb") as f:
        content = f.read()

    original_content = content
    patches_applied = 0

    if b"BuddyBleTransport" not in content:
        print("  [SKIP] No BuddyBleTransport references (feature not present)")
        return True

    # ── Patch A: Force the Buddy feature flag on Linux ────────────────
    #
    # Upstream:  const <ID>="2358734848",<FN>=()=><READER>(<ID>)
    # Patched:   ...,<FN>=()=>process.platform==="linux"||<READER>(<ID>)

    already_a = b'process.platform==="linux"||' in content and b'"2358734848"' in content
    if already_a:
        print("  [OK] Buddy flag: already patched (skipped)")
        patches_applied += 1
    else:
        flag_pattern = (
            rb'(const [\w$]+="2358734848",)'
            rb"([\w$]+=\(\)=>)"
            rb"([\w$]+\([\w$]+\))"
        )

        def flag_replacement(m):
            return m.group(1) + m.group(2) + b'process.platform==="linux"||' + m.group(3)

        content, count = re.subn(flag_pattern, flag_replacement, content, count=1)
        if count >= 1:
            print(f"  [OK] Buddy flag: forced ON for Linux ({count} match)")
            patches_applied += 1
        else:
            print("  [FAIL] Buddy flag pattern not found")
            return False

    # ── Patch B: Early IPC stubs to prevent race condition ────────────
    #
    # The buddy_window renderer calls reportState("ready") during preload,
    # before the async BLE bridge init (setImplementation) completes.
    # Register no-op handlers early; they're replaced when the real bridge
    # calls setImplementation() (which does removeHandler then handle).

    already_b = b'_bi=require("electron").ipcMain' in content
    if already_b:
        print("  [OK] Early stubs: already patched (skipped)")
        patches_applied += 1
    else:
        uuid = extract_eipc_uuid(content)
        if not uuid:
            print("  [FAIL] Could not extract eipc UUID")
            return False

        eipc_prefix = f"$eipc_message$_{uuid}_$_"
        print(f"  [OK] Extracted eipc UUID: {uuid}")

        stub_js = build_early_stubs_js(eipc_prefix)

        pattern = rb'(app\.on\("ready",async\(\)=>\{)'
        replacement = rb"\1" + stub_js.encode("utf-8")

        content, count = re.subn(pattern, replacement, content, count=1)
        if count >= 1:
            print(f"  [OK] Early stubs: injected at app.on('ready') ({count} match)")
            patches_applied += 1
        else:
            print('  [FAIL] app.on("ready") pattern: 0 matches')
            return False

    if patches_applied < EXPECTED_PATCHES:
        print(f"  [FAIL] Only {patches_applied}/{EXPECTED_PATCHES} patches applied")
        return False

    if content != original_content:
        with open(filepath, "wb") as f:
            f.write(content)
        print("  [PASS] Buddy BLE enabled on Linux (flag + early stubs)")
        return True
    else:
        print("  [OK] Already patched, no changes needed")
        return True


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <path_to_index.js>")
        sys.exit(1)

    success = patch_buddy_ble(sys.argv[1])
    sys.exit(0 if success else 1)

File: patches/fix_buddy_ble_linux.nim
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: nim
# Enable Hardware Buddy (Nibblet BLE device) on Linux.
# Two patches: force GrowthBook flag and early IPC stubs.

import std/[os, strformat, strutils]
import regex

const EXPECTED_PATCHES = 2

proc replaceFirst(
    content: var string, pattern: Regex2, subFn: proc(m: RegexMatch2, s: string): string
): int =
  var found = false
  var resultStr = ""
  var lastEnd = 0
  for m in content.findAll(pattern):
    if not found:
      let bounds = m.boundaries
      resultStr &= content[lastEnd ..< bounds.a]
      resultStr &= subFn(m, content)
      lastEnd = bounds.b + 1
      found = true
      break
  if found:
    resultStr &= content[lastEnd .. ^1]
    content = resultStr
    return 1
  return 0

proc extractEipcUuid(content: string): string =
  let pat =
    re2"\$eipc_message\$_([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})"
  for m in content.findAll(pat):
    return content[m.group(0)]
  return ""

proc buildEarlyStubsJs(eipcPrefix: string): string =
  let buddy = eipcPrefix & "claude.buddy_$_Buddy_$_"
  let ble = eipcPrefix & "claude.buddy_$_BuddyBleTransport_$_"
  result =
    "if(process.platform===\"linux\"){" & "const _bi=require(\"electron\").ipcMain;" &
    "_bi.handle(\"" & ble & "reportState\",()=>{});" & "_bi.handle(\"" & ble &
    "rx\",()=>{});" & "_bi.handle(\"" & ble & "log\",()=>{});" & "_bi.handle(\"" & buddy &
    "status\",()=>({connected:false,error:null,paired:false}));" & "_bi.handle(\"" &
    buddy & "deviceStatus\",()=>null);" & "_bi.handle(\"" & buddy &
    "setName\",()=>null);" & "_bi.handle(\"" & buddy & "pairDevice\",()=>\"\");" &
    "_bi.handle(\"" & buddy & "scanDevices\",()=>[]);" & "_bi.handle(\"" & buddy &
    "pickDevice\",()=>false);" & "_bi.handle(\"" & buddy & "cancelScan\",()=>{});" &
    "_bi.handle(\"" & buddy & "forgetDevice\",()=>{});" & "_bi.handle(\"" & buddy &
    "pickFolder\",()=>null);" & "_bi.handle(\"" & buddy & "preview\",()=>null);" &
    "_bi.handle(\"" & buddy & "install\",()=>{});" & "}"

proc apply*(input: string): string =
  result = input
  var patchesApplied = 0

  if "BuddyBleTransport" notin result:
    echo "  [SKIP] No BuddyBleTransport references (feature not present)"
    return result

  # Patch A: Force the Buddy feature flag on Linux
  let alreadyA =
    "process.platform===\"linux\"||" in result and "\"2358734848\"" in result
  if alreadyA:
    echo "  [OK] Buddy flag: already patched (skipped)"
    patchesApplied += 1
  else:
    let flagPattern =
      re2"(const [\w$]+=""2358734848"",)([\w$]+=\(\)=>)([\w$]+\([\w$]+\))"
    var countFlag = result.replaceFirst(
      flagPattern,
      proc(m: RegexMatch2, s: string): string =
        s[m.group(0)] & s[m.group(1)] & "process.platform===\"linux\"||" & s[m.group(2)],
    )
    if countFlag >= 1:
      echo &"  [OK] Buddy flag: forced ON for Linux ({countFlag} match)"
      patchesApplied += 1
    else:
      echo "  [FAIL] Buddy flag pattern not found"
      raise
        newException(ValueError, "fix_buddy_ble_linux: Buddy flag pattern not found")

  # Patch B: Early IPC stubs to prevent race condition
  let alreadyB = "_bi=require(\"electron\").ipcMain" in result
  if alreadyB:
    echo "  [OK] Early stubs: already patched (skipped)"
    patchesApplied += 1
  else:
    let uuid = extractEipcUuid(result)
    if uuid == "":
      echo "  [FAIL] Could not extract eipc UUID"
      raise newException(ValueError, "fix_buddy_ble_linux: Could not extract eipc UUID")

    let eipcPrefix = "$eipc_message$_" & uuid & "_$_"
    echo &"  [OK] Extracted eipc UUID: {uuid}"

    let stubJs = buildEarlyStubsJs(eipcPrefix)

    let pattern = re2"(app\.on\(""ready"",async\(\)=>\{)"
    var countStub = result.replaceFirst(
      pattern,
      proc(m: RegexMatch2, s: string): string =
        s[m.group(0)] & stubJs,
    )
    if countStub >= 1:
      echo &"  [OK] Early stubs: injected at app.on('ready') ({countStub} match)"
      patchesApplied += 1
    else:
      echo "  [FAIL] app.on(\"ready\") pattern: 0 matches"
      raise newException(
        ValueError, "fix_buddy_ble_linux: app.on(\"ready\") pattern not found"
      )

  if patchesApplied < EXPECTED_PATCHES:
    raise newException(
      ValueError,
      &"fix_buddy_ble_linux: Only {patchesApplied}/{EXPECTED_PATCHES} patches applied",
    )

when isMainModule:
  if paramCount() != 1:
    echo "Usage: fix_buddy_ble_linux <file>"
    quit(1)
  let file = paramStr(1)
  echo "=== Patch: fix_buddy_ble_linux ==="
  echo &"  Target: {file}"
  if not fileExists(file):
    echo &"  [FAIL] File not found: {file}"
    quit(1)
  let input = readFile(file)
  let output = apply(input)
  if output != input:
    writeFile(file, output)
    echo "  [PASS] Buddy BLE enabled on Linux (flag + early stubs)"
  else:
    echo "  [OK] Already patched, no changes needed"
